File: src/main/java/theblockbox/huntersdream/api/interfaces/transformation/ITransformation.java
package theblockbox.huntersdream.api.interfaces.transformation;

import net.minecraft.entity.EntityLivingBase;
import net.minecraft.nbt.NBTTagCompound;
import theblockbox.huntersdream.api.Transformation;
import theblockbox.huntersdream.api.helpers.TransformationHelper;
import theblockbox.huntersdream.util.exceptions.WrongTransformationException;

/**
 * This interface is for entites that can transform
 */
public interface ITransformation {
    public Transformation getTransformation();

    public void setTransformation(Transformation transformation);

    public int getTextureIndex();

    public void setTextureIndex(int index);

    /**
     * Sets the texture index of this capability to the texture returned from
     * {@link Transformation#getTextureIndexForEntity(EntityLivingBase)}. The entity
     * should be the entity to which this capability belongs, otherwise the texture
     * index won't be chosen properly and method may even throw a
     * {@link WrongTransformationException}.
     */
    public default void setTextureIndex(EntityLivingBase entity) {
        this.setTextureIndex(this.getTransformation().getTextureIndexForEntity(entity));
    }

    public default boolean textureIndexInBounds() {
        int textureLength = this.getTransformation().getTextures().length;
        // if the texture length is 0, the texture index won't be used and we can return
        // true because nothing needs to be changed
        return ((textureLength == 0) || (this.getTextureIndex() >= 0 && this.getTextureIndex() < textureLength));
    }

    /**
     * Returns an NBTTagCompound with transformation specific data (made so that
     * there doesn't have to be a new capability for each transformation). When the
     * entity changes the transformation the NBTTagCompound always has the key
     * "transformation" with the value of the string returned by
     * {@link Transformation#toString()} where the transformation is the entity's
     * new transformation. Prefer to call
     * {@link TransformationHelper#getTransformationData(EntityLivingBase)} instead
     * of this method to avoid crashes when loading old worlds
     */
    public NBTTagCompound getTransformationData();

    /**
     * Sets the transformation data to the given NBTTagCompound
     */
    public void setTransformationData(NBTTagCompound transformationData);

    /**
     * If the entity can change transformation. It's recommended to prefer
     * {@link TransformationHelper#canChangeTransformation(EntityLivingBase)} or
     * {@link TransformationHelper#canChangeTransformationOnInfection(EntityLivingBase)}
     * over this method. This method is NOT used to determine if an entity can
     * transform, but if its transformation can change
     *
     * @return Returns true if the transformation is changeable without accounting
     * for infection
     */
    public default boolean isTransformationChangeable() {
        return this.getTransformation().isTransformation() && (this.getTransformation() == Transformation.HUMAN
                || this.getTransformation() == Transformation.WEREWOLF);
    }
}

File: src/main/kotlin/theblockbox/huntersnightmare/api/transformation/ITransformation.kt
package theblockbox.huntersnightmare.api.transformation

import net.minecraft.nbt.CompoundNBT
import net.minecraft.nbt.INBT
import net.minecraft.util.Direction
import net.minecraftforge.common.capabilities.Capability
import theblockbox.huntersnightmare.api.init.TransformationInit


interface ITransformation {
    var transformation: Transformation
    var transformationData: CompoundNBT

    /**
     * Returns true if this entity can change its [Transformation] from the current one to the one given.
     */
    fun canChangeTransformation(newTransformation: Transformation) : Boolean

    class TransformationImpl : ITransformation {
        override var transformation: Transformation = TransformationInit.human.get()
        override var transformationData: CompoundNBT = CompoundNBT()

        override fun canChangeTransformation(newTransformation: Transformation) = true
    }

    object TransformationStorage : Capability.IStorage<ITransformation> {
        const val TRANSFORMATION = "transformation"
        const val TRANSFORMATION_DATA = "transformationData"

        override fun writeNBT(capability: Capability<ITransformation>, instance: ITransformation, side: Direction?): INBT? {
            val compound = CompoundNBT()
            compound.putString(TRANSFORMATION, instance.transformation.registryName.toString())
            compound.put(TRANSFORMATION_DATA, instance.transformationData)
            return compound
        }

        override fun readNBT(capability: Capability<ITransformation>, instance: ITransformation, side: Direction?, nbt: INBT?) {
            if (nbt is CompoundNBT) {
                instance.transformation = Transformation.getFromName(nbt.getString(TRANSFORMATION))
                val transformationData = nbt.get(TRANSFORMATION_DATA)
                instance.transformationData = if(transformationData is CompoundNBT) transformationData else CompoundNBT()
            }
        }
    }
}